File: src/nylon/objects/NylonFunction.java
package nylon.objects;

import nylon.exceptions.InvalidActionException;
import nylon.exceptions.NylonRuntimeException;
import nylon.exceptions.UnconvertableTypeException;

import java.util.Iterator;

/**
 * Created by Aedan Smith.
 */

public abstract class NylonFunction implements NylonObject {

    @SuppressWarnings("unchecked")
    public NylonStack apply(NylonStack superStack) throws NylonRuntimeException {
        if (superStack.size() != 0 && superStack.peek().getClass() == FunctionSkipObject.class) {
            superStack.pop();
            return new NylonStack();
        }

        NylonStack functionStack = new NylonStack();
        this.applyImpl(superStack, functionStack);

        NylonStack returnStack = new NylonStack();
        if (functionStack.size() != 0) {
            Class clazz = functionStack.peek().getClass();
            NylonObject object;
            while (functionStack.size() != 0) {
                object = functionStack.pop();
                if (object.getClass() == clazz)
                    returnStack.insertElementAt(object, 0);
                else
                    break;
            }
        }
        return returnStack;
    }

    protected abstract void applyImpl(NylonStack args, NylonStack returnStack) throws NylonRuntimeException;

    @Override
    public String toString() {
        return this.getClass().getSimpleName();
    }

    @Override
    public char toChar() throws NylonRuntimeException {
        throw new UnconvertableTypeException(this, Character.class);
    }

    @Override
    public int toInteger() throws NylonRuntimeException {
        throw new UnconvertableTypeException(this, Integer.class);
    }

    @Override
    public double toDouble() throws NylonRuntimeException {
        throw new UnconvertableTypeException(this, Double.class);
    }

    @Override
    public boolean toBoolean() throws NylonRuntimeException {
        throw new UnconvertableTypeException(this, Boolean.class);
    }

    @Override
    public NylonString increment() throws NylonRuntimeException {
        throw new InvalidActionException("increment", this);
    }

    @Override
    public NylonString decrement() throws NylonRuntimeException {
        throw new InvalidActionException("decrement", this);
    }

    @Override
    public NylonObject concatenate(NylonObject nylonObject) throws NylonRuntimeException {
        if (nylonObject instanceof NylonFunction){
            return new NylonFunction() {
                @Override
                protected void applyImpl(NylonStack args, NylonStack returnStack)
                        throws NylonRuntimeException {
                    NylonFunction.this.applyImpl(args, returnStack);
                    returnStack.addAll(((NylonFunction) nylonObject).apply(returnStack));
                }

                @Override
                public String toString() {
                    return "ConcatenatedFunction[" + NylonFunction.this.toString() + ", " + nylonObject.toString() + "]";
                }
            };
        }
        throw new InvalidActionException("concatenate", this);
    }

    @Override
    public NylonFunction clone() {
        try {
            return (NylonFunction) super.clone();
        } catch (CloneNotSupportedException e) {
            e.printStackTrace();
            System.exit(1);
            return null;
        }
    }

    @Override
    public Iterator<NylonObject> iterator(NylonStack nylonStack) throws NylonRuntimeException {
        return apply(nylonStack).iterator();
    }

    @Override
    public Iterator<NylonObject> reverseIterator(NylonStack nylonStack) throws NylonRuntimeException {
        return apply(nylonStack).reverseIterator(nylonStack);
    }

    @Override
    public int compareTo(NylonObject object, NylonStack nylonStack) throws NylonRuntimeException {
        throw new InvalidActionException("compare", this);
    }
}

File: src/nylon/NylonFunction.kt
package nylon

import java.util.*
import java.util.function.Supplier

/**
 * Created by Aedan Smith.
 */

abstract class NylonFunction(val string: String, var argNum: Int = 0) {
    protected var args: ArrayList<NylonFunction> = ArrayList(argNum)

    protected abstract fun applyImpl(stack: NylonStack, args: ArrayList<NylonFunction>)

    fun apply(stack: NylonStack, args: ArrayList<NylonFunction>) {
        return try {
            applyImpl(stack, args)
        } catch (e: Exception) {
            throw NylonException(e, this)
        }
    }

    fun apply(stack: NylonStack) = apply(stack, args)

    fun apply(): NylonStack {
        val stack = NylonStack()
        apply(stack)
        return stack
    }

    fun resolveArgs(argumentSupplier: Supplier<NylonFunction>) {
        while (args.size < argNum) {
            args.add(argumentSupplier.get())
        }
    }

    fun resolveNestedArgs(argumentSupplier: Supplier<NylonFunction>) {
        resolveArgs(argumentSupplier)
        args.forEach { it.resolveNestedArgs(argumentSupplier) }
    }

    fun clone(): NylonFunction {
        return object : NylonFunction(string, argNum) {
            init {
                this@NylonFunction.args.forEach { this.args.add(it.clone()) }
            }

            override fun applyImpl(stack: NylonStack, args: ArrayList<NylonFunction>) {
                this@NylonFunction.applyImpl(stack, args)
            }

            override fun toString(args: ArrayList<NylonFunction>): String {
                return this@NylonFunction.toString(args)
            }
        }
    }

    open fun toString(args: ArrayList<NylonFunction>): String {
        var s = string
        args.forEach { s += "$it" }
        return s
    }

    override fun toString(): String {
        return toString(args)
    }
}

fun createProvider(nylonObject: NylonObject<*>, string: String): NylonFunction {
    return object : NylonFunction(string) {
        override fun applyImpl(stack: NylonStack, args: ArrayList<NylonFunction>) {
            stack.push(nylonObject.clone())
        }
    }
}

fun concatenate(nylonFunctions: ArrayList<NylonFunction>): NylonFunction {
    return object : NylonFunction("") {
        init {
            args = nylonFunctions
        }

        override fun applyImpl(stack: NylonStack, args: ArrayList<NylonFunction>) {
            args.forEach { it.apply(stack) }
        }

        override fun toString(args: ArrayList<NylonFunction>): String {
            var s = string
            args.forEach { s += " "; s += it }
            return "{" + (if (s.isNotEmpty()) s.substring(1) else s) + "}"
        }
    }
}

fun emptyFunction(): NylonFunction = object : NylonFunction("{}") {
    override fun applyImpl(stack: NylonStack, args: ArrayList<NylonFunction>) {}
}
